### ISSUE TITLE
Submap has a faulty assumption when given coordinates in a different frame

### ISSUE THREAD

**Issue by ayshih** [sunpy/sunpy#4896]
`GenericMap.submap()` has a faulty assumption when given coordinates in a different frame: it assumes that the bounding box defined by the four corners of the input quadrangle include all other points within that quadrangle.  Here's an illustrative picture of the problem.  For an AIA map in helioprojective, if we define the quadrangle shown in blue using heliographic Stonyhurst coordinates, the `submap()` call will return the region shown in red.
![download](https://user-images.githubusercontent.com/991759/105371498-abe36b80-5bd2-11eb-9a4e-7c52e820ac90.png)

**Comment by ayshih** [comment]
```python
import matplotlib.pyplot as plt

from astropy.coordinates import SkyCoord
import astropy.units as u

import sunpy.map
from sunpy.coordinates.utils import get_rectangle_coordinates
from sunpy.data.sample import AIA_171_IMAGE

m = sunpy.map.Map(AIA_171_IMAGE)

bottom_left = SkyCoord(40*u.deg, -45*u.deg, frame='heliographic_stonyhurst', obstime=m.date)
sm = m.submap(bottom_left, width=10*u.deg, height=90*u.deg)

fig = plt.figure(figsize=(8, 8))
ax = fig.add_subplot(111, projection=m)

sm.plot()
```
![download (1)](https://user-images.githubusercontent.com/991759/105372177-52c80780-5bd3-11eb-888e-e54344cdefcc.png)

**Comment by Cadair** [comment]
I would agree we documented the current algorithm correctly but ideally I think it would return the bounding box containing the lines of constant lon/lat defined by the corners.

**Comment by jeffreypaul15** [comment]
Is this a requirement? I suppose this is possible through astropy's ```Quadrangle``` and ```Numpy```?

**Comment by Cadair** [comment]
From an out of band discussion with @ayshih about looking at implementing this in `NDCube.crop`:

I think there are probably 2 main algorithmic approaches here:

1) Generate the edges of the bounding box in world space at some sampling resolution, transform these to pixel space, select the min/max pixel indices which give a rectangular region in pixel space containing the whole ROI (in pixel space).
2) Convert all the pixel coordinates to world coordinates, perform a logical operation on the world coordinates calculating which ones fall inside the ROI, generate a pixel bounding box which encompasses all points.

The second one has the major advantage that it will *always* be correct irrespective of the transform (distortions, lookup tables, non-monotonic coordinates etc.), however, it has the major *disadvantage* that you need to do a pixel to world transform for all pixels. As we [already know](https://github.com/sunpy/ndcube/issues/420) in NDCube doing this, even in a memory efficient manner, is computationally expensive.

The first one has the draw back that you need to sample the world coordinates at some resolution, but the advantage that unless that resolution needs to be insanely high to account for the shape of your world coordinates then it's probably a hell of a lot faster.

**Comment by dstansby** [comment]
How about:

3. Reproject the data to a grid of constant lines of coordinates (ie. a 1:1 mapping from axis > coordinate). Index this cube. Then reproject back to the original coordinate system.

This has the advantage that you do truly select only data inside the desired coordinates, and not all data within the maximum value extents bounding box, but would probably be very computationally expensive.

**Comment by Cadair** [comment]
A valid option, thought you would likely still end up with a box which is larger than the world coordinates after the transform back to the source pixel space. You could achieve the same ends by masking the array in the middle of method 2.
